src/main/scala/org/hammerlab/guacamole/readsets/rdd/PartitionedRegions.scala

```diff
@@ -1,98 +1,214 @@
+import org.hammerlab.guacamole.loci.set.LociSet
 import org.hammerlab.guacamole.logging.LoggingUtils.progress
-import org.hammerlab.guacamole.readsets.PerSample
+import org.hammerlab.guacamole.readsets.{NumSamples, PerSample, SampleId}
 import org.hammerlab.guacamole.reference.ReferenceRegion
-import org.hammerlab.guacamole.windowing.SlidingWindow
+import org.hammerlab.magic.accumulables.{HistogramParam, HashMap => MagicHashMap}
+import org.hammerlab.magic.rdd.KeyPartitioner
+import org.hammerlab.magic.stats.Stats
 
 import scala.reflect.ClassTag
 
 /**
- * This object encapsulates functionality for partitioning reference-mapped regions for processing as "pileups".
+ * Groups a [[LociPartitioning]] with an [[RDD[ReferenceRegion]]] that has already been partitioned according to that
+ * partitioning.
  *
- * Given an RDD of regions, it:
- *
- *   - partitions them by genomic loci,
- *   - makes multiple copies of regions that straddle partition boundaries.
- *
- * The result is that, given knowledge of which ranges of genomic loci correspond to which partition of the resulting
- * RDD, downstream code can iterate over the partitions emitted here and build full pileups for the loci that each
- * partition corresponds to.
- *
- * [[WindowFlatMapUtils]] calls this, and uses [[SlidingWindow]] to iterate over the partitioned regions in a manner
- * suitable for pileup-construction by [[PileupFlatMapUtils]].
+ * This means some regions will occur multiple times in the RDD (due to regions straddling partition boundaries, so it's
+ * important not to confuse this with a regular RDD[ReferenceRegion].
  */
+class PartitionedRegions[R <: ReferenceRegion: ClassTag](val numSamples: NumSamples,
+                                                         @transient keyedRegions: RDD[(SampleId, R)],
+                                                         @transient partitioning: LociPartitioning)
+  extends Serializable {
+
+  assert(
+    keyedRegions.getNumPartitions == lociSetsRDD.getNumPartitions,
+    s"reads partitions: ${keyedRegions.getNumPartitions}, loci partitions: ${lociSetsRDD.getNumPartitions}"
+  )
+
+  def sc: SparkContext = keyedRegions.sparkContext
+
+  val numLoci = sc.accumulator(0L, "numLoci")
+
+  def mapPartitions[V: ClassTag](f: (Iterator[(SampleId, R)], LociSet) => Iterator[V]): RDD[V] =
+    keyedRegions
+      .zipPartitions(
+        lociSetsRDD,
+        preservesPartitioning = true
+      )(
+        (regionsIter, lociIter) => {
+          val loci = lociIter.next()
+          if (lociIter.hasNext) {
+            throw new Exception(s"Expected 1 LociSet, found ${1 + lociIter.size}.\n$loci")
+          }
+
+          numLoci += loci.count
+
+          f(regionsIter, loci)
+        }
+      )
+
+  private lazy val partitionLociSets: Array[LociSet] =
+    partitioning
+      .inverse
+      .toArray
+      .sortBy(_._1)
+      .map(_._2)
+
+  lazy val lociSetsRDD: RDD[LociSet] =
```

Comment:
I'm confused that we're putting lociSets into an RDD. A comment here explaining why would be helpful